File: pirates-ios/PiratesGame/Infrastructure/ImageGenerators/FilledRectGenerator.swift
//
//  FilledRectGenerator.swift
//  Image generators: generates a rectangle filled with a color
//

import UIKit
import ViewletCreator

class FilledRectGenerator: UIImageGenerator {

    func generate(color: UIColor, size: CGSize? = nil, cornerRadius: CGFloat = 0, horizontalGravity: CGFloat = 0.5, verticalGravity: CGFloat = 0.5, forceImageSize: CGSize? = nil, onImage: UIImage? = nil) -> UIImage? {
        if let drawing = beginImageDrawing(withSize: size, horizontalGravity: horizontalGravity, verticalGravity: verticalGravity, forceImageSize: forceImageSize, onImage: onImage) {
            // Cut out shape to allow transparent overdraw without blending
            let rect = drawing.drawRect
            if !drawing.cleanStart && color.cgColor.alpha < 1 {
                // Determine the rectangle, shrink shape if it's not on a pixel boundary to improve edge blending
                var cutRect = CGRect(x: rect.origin.x, y: rect.origin.y, width: rect.width, height: rect.height)
                if abs(cutRect.origin.x - floor(cutRect.origin.x)) > 0.001 {
                    cutRect.origin.x += 0.5
                    cutRect.size.width -= 0.5
                }
                if abs(cutRect.origin.y - floor(cutRect.origin.y)) > 0.001 {
                    cutRect.origin.y += 0.5
                    cutRect.size.height -= 0.5
                }
                if abs(cutRect.origin.x + cutRect.size.width - floor(cutRect.origin.x + cutRect.size.width)) > 0.001 {
                    cutRect.size.width -= 0.5
                }
                if abs(cutRect.origin.y + cutRect.size.height - floor(cutRect.origin.y + cutRect.size.height)) > 0.001 {
                    cutRect.size.height -= 0.5
                }
                
                // Clear shape
                drawing.context.setBlendMode(.clear)
                if cornerRadius > 0 {
                    let path = UIBezierPath(roundedRect: cutRect, cornerRadius: cornerRadius)
                    drawing.context.saveGState()
                    drawing.context.addPath(path.cgPath)
                    drawing.context.fillPath()
                    drawing.context.restoreGState()
                } else {
                    drawing.context.fill(cutRect)
                }
                drawing.context.setBlendMode(.normal)
            }
            
            // Draw
            drawing.context.setFillColor(color.cgColor)
            if cornerRadius > 0 {
                let path = UIBezierPath(roundedRect: rect, cornerRadius: cornerRadius)
                drawing.context.addPath(path.cgPath)
                drawing.context.fillPath()
            } else {
                drawing.context.fill(rect)
            }
            return endDrawingResult()
        }
        return nil
    }
    
    override func generate(attributes: [String: Any], onImage: UIImage? = nil) -> UIImage? {
        let gravity = gravityFromAttributes(attributes)
        return generate(color: ViewletConvUtil.asColor(value: attributes["color"]) ?? UIColor.clear, size: sizeFromAttributes(attributes), cornerRadius: ViewletConvUtil.asDimension(value: attributes["cornerRadius"]) ?? 0, horizontalGravity: gravity.x, verticalGravity: gravity.y, forceImageSize: imageSizeFromAttributes(attributes), onImage: onImage)
    }

}

File: pirates-android/app/src/main/java/com/crescentflare/piratesgame/infrastructure/imagegenerators/FilledRectGenerator.kt
package com.crescentflare.piratesgame.infrastructure.imagegenerators

import android.content.Context
import android.graphics.Color
import android.graphics.PorterDuff
import android.graphics.PorterDuffXfermode
import android.graphics.RectF
import android.graphics.drawable.Drawable
import com.crescentflare.viewletcreator.utility.ViewletMapUtil

/**
 * Image generators: a base class to generate a new image drawable dynamically
 */
class FilledRectGenerator: ImageDrawableGenerator() {

    fun generate(context: Context, color: Int, width: Int? = null, height: Int? = null, cornerRadius: Int = 0, horizontalGravity: Float = 0.5f, verticalGravity: Float = 0.5f, forceImageWidth: Int? = null, forceImageHeight: Int? = null, onDrawable: Drawable? = null): Drawable? {
        val drawing = beginImageDrawing(width, height, horizontalGravity, verticalGravity, forceImageWidth, forceImageHeight, onDrawable)
        if (drawing != null) {
            // Cut out shape to allow transparent overdraw without blending
            val rect = drawing.drawRect
            if (!drawing.cleanStart && (color.toLong() and 0xff000000) != 0xff000000) {
                // Determine the rectangle, shrink shape if it's not on a pixel boundary to improve edge blending
                val cutRect = RectF(rect.left, rect.top, rect.right, rect.bottom)
                if (Math.abs(cutRect.left - Math.floor(cutRect.left.toDouble())) > 0.001f) {
                    cutRect.left += 0.5f
                }
                if (Math.abs(cutRect.top - Math.floor(cutRect.top.toDouble())) > 0.001f) {
                    cutRect.top += 0.5f
                }
                if (Math.abs(cutRect.right - Math.floor(cutRect.right.toDouble())) > 0.001f) {
                    cutRect.right += 0.5f
                }
                if (Math.abs(cutRect.bottom - Math.floor(cutRect.bottom.toDouble())) > 0.001f) {
                    cutRect.bottom += 0.5f
                }

                // Clear shape
                drawing.paint.color = Color.TRANSPARENT
                drawing.paint.xfermode = PorterDuffXfermode(PorterDuff.Mode.CLEAR)
                if (cornerRadius > 0) {
                    drawing.canvas.drawRoundRect(cutRect, cornerRadius.toFloat(), cornerRadius.toFloat(), drawing.paint)
                } else {
                    drawing.canvas.drawRect(cutRect, drawing.paint)
                }
                drawing.paint.xfermode = null
            }

            // Draw shape
            drawing.paint.color = color
            if (cornerRadius > 0) {
                drawing.canvas.drawRoundRect(rect, cornerRadius.toFloat(), cornerRadius.toFloat(), drawing.paint)
            } else {
                drawing.canvas.drawRect(rect.left, rect.top, rect.right, rect.bottom, drawing.paint)
            }
            return endDrawingResult(context, drawing)
        }
        return null
    }

    override fun generate(context: Context, attributes: Map<String, Any>, onDrawable: Drawable?): Drawable? {
        val gravity = gravityFromAttributes(attributes)
        return generate(
            context,
            ViewletMapUtil.optionalColor(attributes, "color", Color.TRANSPARENT),
            widthFromAttributes(attributes),
            heightFromAttributes(attributes),
            ViewletMapUtil.optionalDimension(attributes, "cornerRadius", 0),
            gravity.x,
            gravity.y,
            imageWidthFromAttributes(attributes),
            imageHeightFromAttributes(attributes),
            onDrawable
        )
    }

}
